knack/deprecation.py

```diff
@@ -0,0 +1,188 @@
+
+
+# pylint: disable=too-many-instance-attributes
+class Deprecated(object):
+
+    @staticmethod
+    def ensure_new_style_deprecation(cli_ctx, kwargs, object_type):
+        """ Helper method to make the previous string-based deprecate_info kwarg
+            work with the new style. """
+        deprecate_info = kwargs.get('deprecate_info', None)
+        if isinstance(deprecate_info, Deprecated):
+            deprecate_info.object_type = object_type
+        elif isinstance(deprecate_info, STRING_TYPES):
+            deprecate_info = Deprecated(cli_ctx, redirect=deprecate_info, object_type=object_type)
+        kwargs['deprecate_info'] = deprecate_info
+        return deprecate_info
+
+    def __init__(self, cli_ctx=None, object_type='', target=None, redirect=None, hide=False, expiration=None,
+                 tag_func=None, message_func=None):
+        """ Create a collection of deprecation metadata.
+
+        :param cli_ctx: The CLI context associated with the deprecated item.
+        :type cli_ctx: knack.cli.CLI
+        :param object_type: A label describing the type of object being deprecated.
+        :type: object_type: str
+        :param target: The name of the object being deprecated.
+        :type target: str
+        :param redirect: The alternative to redirect users to in lieu of the deprecated item. If omitted it, there is
+                         no alternative.
+        :type redirect: str
+        :param hide: A boolean or CLI version at or above-which the deprecated item will no longer appear
+                     in help text, but will continue to work. Warnings will be displayed if the deprecated
+                     item is used.
+        :type hide: bool OR str
+        :param expiration: The CLI version at or above-which the deprecated item will no longer work.
+        :type expiration: str
+        :param tag_func: Callable which returns the desired unformatted tag string for the deprecated item.
+                         Omit to use the default.
+        :type tag_func: callable
+        :param message_func: Callable which returns the desired unformatted message string for the deprecated item.
+                             Omit to use the default.
+        :type message_func: callable
+        """
+        self.cli_ctx = cli_ctx
+        self.object_type = object_type
+        self.target = target
+        self.redirect = redirect
+        self.hide = hide
+        self.expiration = expiration
+
+        def _default_get_message(self):
+            line1 = "This {} has been deprecated and will be removed ".format(self.object_type)
+            if self.expiration:
+                line1 += "in version '{}'.".format(self.expiration)
+            else:
+                line1 += 'in a future release.'
+            lines = [line1]
+            if self.redirect:
+                lines.append("Use '{}' instead.".format(self.redirect))
+            return ' '.join(lines)
```

Comment:
If lines is only going to have 1 or 2 elements, why not just add to the string if self.redirect?

Reply:
TL/DR -> Trust me 😆 

Because you have to have annoying logic to include a space or not, and you end up with a whitespace nightmare. This ensures there is always and only ONE space between lines. I tried other schemes and it was a constant headache getting unit tests to pass due to "off-by-one" errors related exactly to trying to just add to strings.

Reply:
okay I trust you

Reply:
Actually, I did you suggestion and it worked fine.  Having unit tests is a good thing 👍